pkg/agent/clean/binding.go

```diff
@@ -0,0 +1,295 @@
+		return nil
+	}
+
+	var waitGroup sync.WaitGroup
+
+	waitGroup.Add(2)
+	go func() {
+		if err := bc.cleanCRTB(rancher25, crtbs.Items); err != nil {
+			logrus.Error(err)
+		}
+		waitGroup.Done()
+	}()
+
+	go func() {
+		if err := bc.cleanPRTB(rancher25, prtbs.Items); err != nil {
+			logrus.Error(err)
+		}
+		waitGroup.Done()
+	}()
+	waitGroup.Wait()
+	return nil
+}
+
+func (bc *bindingsCleanup) cleanCRTB(newLabel bool, crtbs []apiv3.ClusterRoleTemplateBinding) error {
+	var objectMetas []metav1.ObjectMeta
+	for _, crtb := range crtbs {
+		objectMetas = append(objectMetas, crtb.ObjectMeta)
+	}
+
+	return bc.cleanObjectDuplicates(crtbType, newLabel, objectMetas)
+}
+
+func (bc *bindingsCleanup) cleanPRTB(newLabel bool, prtbs []apiv3.ProjectRoleTemplateBinding) error {
+	var objectMetas []metav1.ObjectMeta
+	for _, prtb := range prtbs {
+		objectMetas = append(objectMetas, prtb.ObjectMeta)
+	}
+
+	return bc.cleanObjectDuplicates(prtbType, newLabel, objectMetas)
+}
+
+func (bc *bindingsCleanup) cleanObjectDuplicates(bindingType string, newLabel bool, objMetas []metav1.ObjectMeta) error {
+	// Uppercase so the logging looks pretty
+	bindingUpper := strings.ToUpper(bindingType)
+
+	var returnErr error
+	var totalCRBDupes, totalRoleDupes int
+
+	for _, meta := range objMetas {
+		labels := createLabelSelectors(newLabel, meta, bindingType)
+		for _, label := range labels {
+			var CRBduplicates, RBDupes int
+
+			crbs, err := bc.clusterRoleBindings.List(metav1.ListOptions{LabelSelector: label})
+			if err != nil {
+				multierror.Append(returnErr, err)
+			}
+
+			if len(crbs.Items) > 1 {
+				CRBduplicates += len(crbs.Items) - 1
+				if err := bc.dedupeCRB(crbs.Items); err != nil {
+					multierror.Append(returnErr, err)
+				}
+			}
+
+			roleBindings, err := bc.roleBindings.List("", metav1.ListOptions{LabelSelector: label})
+			if err != nil {
+				multierror.Append(returnErr, err)
+			}
+
+			if len(roleBindings.Items) > 1 {
+				roleDuplicates, err := bc.dedupeRB(roleBindings.Items)
+				if err != nil {
+					multierror.Append(returnErr, err)
+				}
+				RBDupes += roleDuplicates
+			}
+			if CRBduplicates > 0 || RBDupes > 0 {
+				totalCRBDupes += CRBduplicates
+				totalRoleDupes += RBDupes
+				logrus.Infof("%v %v label:%v Duplicates: CRB:%v RB:%v", bindingUpper, meta.Name, label, CRBduplicates, RBDupes)
+			}
+		}
+	}
+	logrus.Infof("Total %v duplicate clusterRoleBindings %v, roleBindings %v", bindingUpper, totalCRBDupes, totalRoleDupes)
+	return returnErr
+}
+
+func (bc *bindingsCleanup) dedupeCRB(bindings []k8srbacv1.ClusterRoleBinding) error {
+	// Sort by creation timestamp so we keep the oldest
+	sort.Sort(crbByCreation(bindings))
+
+	// Leave the first one alone, we only need the duplicates
+	duplicates := bindings[1:]
+
+	for _, binding := range duplicates {
+		if !dryRun {
+			bc.clusterRoleBindings.Delete(binding.Name, &metav1.DeleteOptions{})
```

Comment:
should we handle/log delete error?

Reply:
Logging the error on both delete calls now.